src/server/game/AI/SmartScripts/SmartScript.cpp

```diff
@@ -1123,22 +1123,39 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
             delete targets;
             break;
         }
-        case SMART_ACTION_MOVE_FORWARD:
+        case SMART_ACTION_MOVE_OFFSET:
         {
-            if (!me)
-                break;
+            if (ObjectList* targets = GetTargets(e, unit))
+            {
+                for (ObjectList::const_iterator itr = targets->begin(); itr != targets->end(); ++itr)
+                {
+                    if (!IsCreature(*itr))
+                        continue;
 
-            float x, y, z;
-            me->GetClosePoint(x, y, z, me->GetObjectSize() / 3, (float)e.action.moveRandom.distance);
-            me->GetMotionMaster()->MovePoint(SMART_RANDOM_POINT, x, y, z);
-            break;
-        }
-        case SMART_ACTION_RISE_UP:
-        {
-            if (!me)
-                break;
+                    Position pos = (*itr)->GetPosition();
+
+                    if (e.action.offsetMovement.movement)
+                    {
+                        // Use normal WoW axis-based movement
+                        Position offset = { e.target.x, e.target.y, e.target.z, 0.0f };
+                        pos.RelocateOffset(offset);
+                        (*itr)->ToCreature()->GetMotionMaster()->MovePoint(SMART_RANDOM_POINT, pos);
+                    }
+                    else
+                    {
+                        // Use forward/backward/left/right cartesian plane movement
+                        float x, y, z, o;
+                        o = pos.GetOrientation();
+                        x = pos.GetPositionX() + (cos(o - (M_PI / 2))*e.target.x) + (cosf(o)*e.target.y);
+                        y = pos.GetPositionY() + (sin(o - (M_PI / 2))*e.target.x) + (sinf(o)*e.target.y);
```

Comment:
This calculates a point directly forward/backward/left/right based on creature's orientation, allowing the creature to walk, say, 10 yards forward no matter what its orientation is. The original RelocateOffset calculation doesn't seem to do that (tested with different values, never got the "expected" result).

The purpose of this is to allow to specify a human-friendly value. Say, if you set target_y to 10, the creature will move 10 yards forward. Think of the cartesian plane, x acts as left/right and y acts as forward/backward

Honestly, I don't quite understand the math behind it. I only know it works, while RelocateOffset gives different values :)